packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts

```diff
@@ -0,0 +1,231 @@
+import { EventEmitter } from 'stream'
+import { AbortError } from 'timeline-state-resolver'
+import { Logger } from 'winston'
+import { Job } from './job'
+
+interface JobDescription<X> {
+	id?: string
+	job: Job<X, any, any>
+	chained: boolean
+	abortController: AbortController
+	timeout?: number
+	didTimeOut?: boolean
+	result?: any
+	importance?: JobImportance
+	onFulfilled?: FulfilledCallback<X>
+	onRejected?: RejectedCallback
+	onTimeout?: TimeoutCallback
+}
+
+export enum JobImportance {
+	HIGH = 'high',
+	LOW = 'low',
+}
+
+type FulfilledCallback<T> = (queue: IJobQueue, result: T) => void
+type RejectedCallback = (reason: any) => void
+type TimeoutCallback = (queue: IJobQueue) => void
+
+interface IChainableJobQueue<X> extends IJobQueue {
+	chain: <T, V>(job: Job<T, X, V>, timeout?: number, importance?: JobImportance, id?: string) => IChainableJobQueue<T>
+	end: (onFulfilled?: FulfilledCallback<X>, onRejected?: RejectedCallback, onTimeout?: TimeoutCallback) => void
+}
+
+export interface IJobQueue {
+	get length(): number
+	clear(): void
+	endsWith(jobId: string): boolean
+	getLastJobIdByImportance(importance: JobImportance): string | undefined
+	enqueue<T, V>(
+		job: Job<T, void, V>,
+		timeout?: number,
+		importance?: JobImportance,
+		id?: string
+	): IChainableJobQueue<T>
+	removeJob(jobId: string): void
+}
+
+export class JobQueue extends EventEmitter implements IJobQueue {
+	private queue: JobDescription<any>[] = []
+	private currentJob: JobDescription<any> | null = null
+	private currentJobChain: JobDescription<any>[] = []
+	private currentJobTimeout: NodeJS.Timeout | null = null
+
+	constructor(private name: string, private logger: Logger) {
+		super()
+	}
+
+	get length(): number {
+		return this.queue.length
+	}
+
+	clear(): void {
+		if (this.currentJob) {
+			this.abortJob(this.currentJob)
+		} else {
```

Comment:
By adding `return` inside the if-statement, the else-branch can be dissolved to decrease the number of nested levels.